x/ibc/testing/chain.go

```diff
@@ -0,0 +1,373 @@
+package testing
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/secp256k1"
+	lite "github.com/tendermint/tendermint/lite2"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	connectiontypes "github.com/cosmos/cosmos-sdk/x/ibc/03-connection/types"
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/04-channel/types"
+	ibctmtypes "github.com/cosmos/cosmos-sdk/x/ibc/07-tendermint/types"
+	commitmenttypes "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment/types"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/24-host"
+	"github.com/cosmos/cosmos-sdk/x/ibc/keeper"
+)
+
+const (
+	// Default params used to create a TM client
+	TrustingPeriod  time.Duration = time.Hour * 24 * 7 * 2
+	UnbondingPeriod time.Duration = time.Hour * 24 * 7 * 3
+	MaxClockDrift   time.Duration = time.Second * 10
+
+	ConnectionVersion = "1.0"
+	ChannelVersion    = "1.0"
+
+	ConnectionIDPrefix = "connectionID"
+	ChannelIDPrefix    = "channelID"
+	PortIDPrefix       = "portID"
+)
+
+// TestChain is a testing struct that wraps a simapp with the last TM Header, the current ABCI
+// header and the validators of the TestChain. It also contains a field called ClientID. This
+// is the clientID that *other* chains use to refer to this TestChain. For simplicity's sake
+// it is also the chainID on the TestChain Header. The SenderAccount is used for delivering
+// transactions through the application state.
+type TestChain struct {
+	t *testing.T
+
+	ClientID      string
+	App           *simapp.SimApp
+	LastHeader    ibctmtypes.Header // header for last block height committed
+	CurrentHeader abci.Header       // header for current block height
+	Querier       sdk.Querier
+
+	Vals    *tmtypes.ValidatorSet
+	Signers []tmtypes.PrivValidator
+
+	senderPrivKey crypto.PrivKey
+	SenderAccount authtypes.AccountI
+
+	// IBC specific helpers
+	Channels    []Channel // track portID/channelID's created for this chain
```

Comment:
I don't fully understand what these helpers are for. Are they for ease of retrieving the channels and connections? ~I don't see that they are being updated thought the handshakes. Should we use pointers instead or a `map [portID+channelID] -> *Channel`? Thoughts~

Reply:
ahh lol, I thought you were using the actual channel struct. Ignore the second part of the comment

Reply:
yea I was worried that naming might be a bit confusing. Maybe I should rename it to `TestChannel` or something to be more explicit that it is quite different an actual IBC channel

Reply:
`TestChannel` is better